### ISSUE TITLE
ACF post preview and revisions completely broken

### ISSUE THREAD

**Issue by tripflex** [AdvancedCustomFields/acf#903]
**Describe the bug**
The internal handling of ACF for revisions and previews seems to be completely broken in numerous ways when using the default WordPress Gutenberg editor.  Any field groups created and attached to a post, either do not show at all when attempting to preview, or if using a workaround to remove the filter, they show values from the last "published" post.

TLDR; **Preview does not work correctly for ACF fields** if using the **_default_** WordPress Gutenburg editor

1.) ACF does not handle revisions correctly, and is using the revision ID when you call `get_field` (see end of issue for details), ultimately causing nothing to show because metadata does not exist on the revision post ID.

2.) Even with revisions disabled, preview does not work correctly which probably stems from the same issue above, and does not show anything on preview either.

**To Reproduce**
1. Create a blank WordPress install
2. Create a new field group, leave the defaults to show when post type is **Post**
3. Add any kind of field (text, repeater, etc)
4. Create a new post
5. Add `[acf field="some_field_value"]` shortcode to output on the page
6. Set any random value in the ACF field created above
7. Publish
8. Edit post, change the ACF field value to something different
9. Click Computer icon on top right and select **Preview in new tab**
10. Preview will load, and either show nothing or show the old value

**Expected behavior**
The value updated when editing to show correctly on the preview

**Basic setup and functional when published:**
![image](https://github.com/AdvancedCustomFields/acf/assets/553732/fc7f6c84-e109-47bb-a6c3-959fc114efb8)
![image](https://github.com/AdvancedCustomFields/acf/assets/553732/54614b50-9f76-4df9-9f7c-a481e56b6978)
![image](https://github.com/AdvancedCustomFields/acf/assets/553732/6e0a760f-1d46-4ce6-b5ad-9c7f9aa737f5)

**Now change the field value:**
![image](https://github.com/AdvancedCustomFields/acf/assets/553732/1c6a8e28-e2a9-4548-84b3-73395754f6fb)

**Click preview in new tab:**
![image](https://github.com/AdvancedCustomFields/acf/assets/553732/ab5a25ba-93e5-47b2-ad90-e4e7c7174b56)

**Nothing shows:**
![image](https://github.com/AdvancedCustomFields/acf/assets/553732/f44477d0-762b-4cda-8a7d-3bab92eaa8e9)

**Version Information:**
 - WordPress 6.4.3
 - PHP 7.4 & 8.1
 - ACF PRO Version 6.2.7
 - Any Browser

**Additional context**

As a WordPress expert and seasoned plugin developer, here's what I found.

The problem exists in `acf_get_valid_post_id` which calls the `acf/validate_post_id` filter, which then is filtered in `acf_revisions::acf_validate_post_id` who then calls `acf_get_post_latest_revision` which returns the revision ID.

The problems stems from the fact that ACF does not store meta on the revision, but for some reason, it's specifically changing Post ID's to the revision ID, ultimately causing the function call to `get_field` to call `get_metadata` using the revision post ID, but the only meta that will ever exist on it is something like this:

`
[ '_acf_changed' => [ 0 => "1" ] ]
`

My assumption is this code was meant for the pre-Gutenburg era, as I also found that when WordPress automatically calls the `autosaves` REST endpoint (every 60 seconds or when you click **Preview in new tab**), ACF does not have any handling in there to pass any values that have been changed, and since they are not one of the default fields `post_title, post_content, excerpt, etc` nothing gets saved, and as such, ACF is trying to pull the value from the revision (or autosave if revisions disabled) that will never have anything on it, as ACF **has specifically set to only allow** `_acf_changed` as the only valid meta for revisions/autosaves.

Basically it seems that preview only works with ACF if you specifically force the site to use the classic editor.  I've tested this on a completely blank install and replicated it numerous times.

Numerous previous issues opened regarding this:
https://github.com/AdvancedCustomFields/acf/issues/411
https://github.com/AdvancedCustomFields/acf/issues/186
https://github.com/AdvancedCustomFields/acf/issues/184

**Comment by lgladdy** [comment]
Hey there,

We're aware of this issue, and there are a few long standing issues in this repo for it.

We have a plan to resolve this issue, but it requires us to remove the legacy meta box and transition to a whole new system for updating ACF fields in the block editor, this is a long term project for us.

We'll keep you updated on our progress over the next several releases of ACF.

**Comment by tripflex** [comment]
> Hey there,
> 
> We're aware of this issue, and there are a few long standing issues in this repo for it.
> 
> We have a plan to resolve this issue, but it requires us to remove the legacy meta box and transition to a whole new system for updating ACF fields in the block editor, this is a long term project for us.
> 
> We'll keep you updated on our progress over the next several releases of ACF.

What can I do to help?  I'm happy to do some development work if you can point me in the right direction.  Thanks!

**Comment by ConnorWedia** [comment]
I am also experiencing an issue similar to the above where if I add some text and let it autosave (without modifying any ACF fields), when I go to revisions and view the autosave, ALL ACF fields get removed. 

Paragraph added for autosave test:
<img width="930" alt="added_paragraph" src="https://github.com/AdvancedCustomFields/acf/assets/145330085/e7a71dc3-cc75-46b1-80ee-9d4ff09e9a96">

Autosave Response result:
![autosave_response](https://github.com/AdvancedCustomFields/acf/assets/145330085/c858c619-c053-4c65-a907-1dd2bf83cd68)

Revision of the autosave, showing the removed ACF fields:
<img width="1304" alt="removed_acf_fields" src="https://github.com/AdvancedCustomFields/acf/assets/145330085/cbb181ca-881c-4d85-9d4d-475ba790d1da">


Is there ANY fix for this please as our content writers are constantly re-writing & re-adding ACF content whenever the post autosaves.

Thanks

**Comment by lgladdy** [comment]
@ConnorWedia Just to clarify, is this a recent change with an ACF or WordPress update, or has this always been the case for you in the block editor?

**Comment by lgladdy** [comment]
Thanks @ConnorWedia - I'll raise it with the team here today and see the updates we made to fix (some) revisions made it worse for autosaves.

**Comment by phil-sola** [comment]
@lgladdy is there any movement on previews and revisions with ACF Fields? 

It works for me always after the first time, but the first time I hit preview after changing some ACF flexible fields, it never displays. Super frustrating!

**Comment by ErikBrendel** [comment]
@ConnorWedia did you find a solution to the problem of all ACF values being cleared on autosaves? We're experiencing the same problem...

**Comment by cooperfellows** [comment]
Adding our voice to this issue. Working with ACF Pro and experiencing these same issues. Starting to become a problem with our content editors. Any roadmap that can be shared around the "immediate fix" you mentioned @lgladdy?

**Comment by lgladdy** [comment]
@cooperfellows We're actively working on a potential long term fix for this at the moment and have a working internal branch, we'll have more to share on which release will contain this shortly once we're confident everything is working correctly.

**Comment by fiskhandlarn** [comment]
@lgladdy any news on this?

**Comment by SiteBoltsAdmin** [comment]
We noticed today that WordPress's "Preview Changes" feature wasn't working correctly for ACF fields; the preview was showing their published values rather than the previewed values.

We tried out several suggested PHP snippets but found that they didn't resolve the issue.

After seeing this suggestion:

    define( 'WP_POST_REVISIONS', 3 );

We did some testing and confirmed that the issue occurs when post revisions are disabled. In our case, it was due to our WordPress host (WPEngine) disabling post revisions by default. We were able to fix it by contacting their support and having them set the number of post revisions to 3.

**Comment by lgladdy** [comment]
For information on testing ACF PRO Betas, you can join the beta mailing list at https://www.advancedcustomfields.com/beta-news/

We'll likely have our next beta at some point next month, but we can't guarantee any timelines at the moment.

Thanks,
Liam

**Comment by fiskhandlarn** [comment]
@lgladdy is this fix currently in the beta?

**Comment by eszpee** [comment]
Hey @lgladdy, based on your September message last year:

> [@cooperfellows](https://github.com/cooperfellows) We're actively working on a potential long term fix for this at the moment and have a working internal branch, we'll have more to share on which release will contain this shortly once we're confident everything is working correctly.

and this one from January:

> We'll likely have our next beta at some point next month, but we can't guarantee any timelines at the moment.

It seems like you were really close to releasing a fix for this issue, which impacts us too. Could you provide an update on what went wrong and when can we expect to have this resolved? 

Thanks a lot in advance!

**Comment by eszpee** [comment]
Thanks for the quick reply, crossing my fingers for you to get the bandwidth for that refactor.

**Comment by VladimirMorozov1994** [comment]
@lgladdy Hello, is there any news on this issue? )

**Comment by marcelpietsch** [comment]
This issue still exists. Any updates?
